app/src/main/java/de/westnordost/streetcomplete/quests/barrier_type/DetectWayBarrierIntersectionUtil.kt

```diff
@@ -0,0 +1,136 @@
+package de.westnordost.streetcomplete.quests.barrier_type
+
+import de.westnordost.streetcomplete.data.elementfilter.ElementFilterExpression
+import de.westnordost.streetcomplete.data.osm.mapdata.Element
+import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
+import de.westnordost.streetcomplete.data.osm.mapdata.MapDataWithGeometry
+import de.westnordost.streetcomplete.data.osm.mapdata.Way
+import de.westnordost.streetcomplete.util.isRightOf
+
```

Comment:
There is a lot of copy-pasta here from the `AddCrossing` quest. You should define this function generically like this

```kotlin
/** Return all nodes that are at crossing points between waysA and waysB.
 *
 *  In this context, two ways cross if
 *  1. they intersect at a shared node
 *  2. the node is not the end node of any of the two ways
 *  3. the ways actually cross each other, not use touch each other, i.e. the way A
 *     intersects, then continues on the other side of way B
 */
fun findNodesAtCrossingsOf(waysA: Iterable<Way>, waysB: Iterable<Way>, mapData: MapData): Iterable<Node>
````

and then make the `AddCrossing` code use that too.

Not sure about the special handling of sidewalk stuff (line 57-69 in `AddCrossing`) and tunnel-bridge stuff. The cleanest solution I can think of is to have the function not return just the node at the crossing, but the node+ways that are part of the crossing so that it can be filtered further.
I.e.
```kotlin
data class WaysCrossing(val node: Node, ways: List<Way>)

fun findNodesAtCrossingsOf(waysA: Iterable<Way>, waysB: Iterable<Way>, mapData: MapData): Iterable<Crossing>
```

Reply:
> you should define this function generically

I tried doing this, but everything felt incredibly generically what made any reasoning about code more complicated.

I will try once more.

Reply:
IMO, some incidental duplication is less bad than coupling things via the wrong abstraction. If it's only duplicated between *two* quests, I wouldn't try too hard to de-duplicate, unless we are very sure that they are *actually* the same exact logic, and not something which just looks similar.

Reply:
The logic to determine if two ways actually cross at a node is both complex and generic enough that it is worth generalizing.

Reply:
A middle ground would be a comment in both places which warns that if you feel a need to copy-paste again, it should be pulled out into a helper function first.

Reply:
> The cleanest solution I can think of is to have the function not return just the node at the crossing, but the node+ways that are part of the crossing so that it can be filtered further.

Is it not going to be performance problematic due to generating and passing even more data? Though likely in both cases number of objects should be limited, without noticeable impact.

I also thought about passing callback function instead - that would allow further filtering without passing this data.

Reply:
> Though likely in both cases number of objects should be limited, without noticeable impact.

yes

Yes, I also thought about callback function instead. But I find this solution quite dirty.

Reply:
`waysA` `waysB` - I think that I will name them `movingWays` and `barrierWays` at least while coding.

Otherwise it is too easy to confuse them with each other with generic `A` `B`.

I can refactor it to more generic names at the end or keep that. Or maybe there would be a better naming?